README.md
[![Waffle.io - Columns and their card count](https://badge.waffle.io/iamthechad/parallax_lcd.svg?columns=all)](https://waffle.io/iamthechad/parallax_lcd)


<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*

- [Parallax LCD](#parallax-lcd)
  - [Functions](#functions)
  - [Example](#example)
  - [Installing](#installing)
    - [Download](#download)
    - [Manual Installation](#manual-installation)
    - [After Installation](#after-installation)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

# Parallax LCD

The Parallax LCD library allows for writing to Parallax three wire LCDs.
It is known to work with Parallax LCD's (27976, 27977, 27979).

For more information about Parallax LCDs, see [http://www.parallax.com](http://www.parallax.com)	

## Functions

The library offers the following functions:

* `setup`
* `on`
* `off`
* `scrollLeft`
* `scrollRight`
* `empty`
* `cursorUnderline`
* `cursorBlock`
* `cursorOff`
* `cursorOnChrBlink`
* `cr`
* `lf`
* `backLightOn`
* `backLightOff`
* `playTone`
* `setCustomCharacter`
* `printCustomCharacter`
* `pos`
* `at`

## Example

Using the library is simple. See the included example for more.

    #include <ParallaxLCD.h>

    ParallaxLCD lcd(2,2,16); // desired pin, rows, cols

    void setup () {
      lcd.setup();
      delay(1000); // Wait for the LCD to start up
      lcd.at(1,3,"Hello World!");
    }

## Installing

Installation is performed via the Arduino Library Manager (Sketch -> Include Library)

### Download

* In the library manager, select "Manage Libraries..."
* In the search box, enter "Parallax LCD"
* Select the library in the results, pick the desired version and click "Install"

### Manual Installation

* Download the latest zip from the Releases tab. (Alternatively - clone this repository or select the "Zip" button above to download the code.)
* In the library manager, select "Add .ZIP Library"
* Select the downloaded zip file

### After Installation
* Once installed, you may need to restart the Arduino IDE
* Go to `File`->`Examples`. There should now be an entry for `Parallax LCD` in the list. Select it to see the example.
* Using the example as a guide, write your own code!

# License

This library is provided under the MIT license. You can obtain a copy of this license at http://opensource.org/licenses/MIT.

CHANGELOG.md
# Change Log

## [1.2](https://github.com/iamthechad/parallax_lcd/tree/1.2) (2018-03-07)
[Full Changelog](https://github.com/iamthechad/parallax_lcd/compare/1.1...1.2)

**Implemented enhancements:**

- Update for Arduino 1.5 format [\#7](https://github.com/iamthechad/parallax_lcd/issues/7)
- Build error if SoftwareSerial.h not included [\#6](https://github.com/iamthechad/parallax_lcd/issues/6)

**Fixed bugs:**

- Build error if SoftwareSerial.h not included [\#6](https://github.com/iamthechad/parallax_lcd/issues/6)

**Merged pull requests:**

- Update for release [\#10](https://github.com/iamthechad/parallax_lcd/pull/10) ([iamthechad](https://github.com/iamthechad))
- Update to Arduino 1.5 format [\#9](https://github.com/iamthechad/parallax_lcd/pull/9) ([iamthechad](https://github.com/iamthechad))
- Add import to header [\#8](https://github.com/iamthechad/parallax_lcd/pull/8) ([iamthechad](https://github.com/iamthechad))

## [1.1](https://github.com/iamthechad/parallax_lcd/tree/1.1) (2015-01-16)
[Full Changelog](https://github.com/iamthechad/parallax_lcd/compare/1.0...1.1)

**Fixed bugs:**

- 4 x 20 LCD does not work correctly [\#5](https://github.com/iamthechad/parallax_lcd/issues/5)

**Merged pull requests:**

- waffle.io Badge [\#4](https://github.com/iamthechad/parallax_lcd/pull/4) ([waffle-iron](https://github.com/waffle-iron))

## [1.0](https://github.com/iamthechad/parallax_lcd/tree/1.0) (2013-12-19)
**Implemented enhancements:**

- Assign an OSS license [\#3](https://github.com/iamthechad/parallax_lcd/issues/3)

**Merged pull requests:**

- Update examples/LCD\_2x16/LCD\_2x16.pde [\#2](https://github.com/iamthechad/parallax_lcd/pull/2) ([mcstoufer](https://github.com/mcstoufer))
- Update ParallaxLCD.cpp [\#1](https://github.com/iamthechad/parallax_lcd/pull/1) ([mcstoufer](https://github.com/mcstoufer))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

src/ParallaxLCD.h
/* 
	NOTE: you must: #include <SoftwareSerial.h>
	BEFORE including the class header file
	
	For Parallax LCD's (27976, 27977, 27979)
	For more information contact http://www.parallax.com	
	
	Updated to Arduino 1.0 by Chad Johnston (cjohnston@megatome.com)
	July 11, 2012
	      
	Original Work done by:
    Douglas Farrell	
    cyberblob59@yahoo.com
			  
    Used/Modified with permission from allen joslin
    --Based on SparkFunSerLCD.h from--
	allen joslin
	payson productions
	allen@joslin.net
	
	Updated 8/1/2012 Martin C. Stoufer mcstoufer@speakeasy.net
	# Added interface functions to play music and define/print custom characters.
*/
/*
The MIT License (MIT)

Copyright (c) 2012-2018 Megatome Technologies, LLC

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#ifndef ParallaxLCD_h
#define ParallaxLCD_h

#include <SoftwareSerial.h>
#include <Arduino.h>

/******************************************************************************************************/
/* ParallaxLCD -- manages the Parallax LCD's, based on NewSoftSerial to aid pinning and printing      */
/*                                                                                                    */
/*	   Stable Serial Speeds 2400, 9600, 19200                                                         */
/*                                                                                                    */
/*     posBase: cursor positioning                                                                    */
/*                                                                                                    */
/*     on/off: display of characters                                                                  */
/*                                                                                                    */
/*     bright: backlight control, on or off                                                           */
/*                                                                                                    */
/*     LF: Line Feed control                                                                          */
/*                                                                                                    */
/*     CR: Carrage Return control                                                                     */
/*                                                                                                    */
/*     Cursor: Cursor Control                                                                         */
/*                                                                                                    */
/*     Music: Generate tones for Piezo speaker                                                        */
/*                                                                                                    */
/*     Custom Characters: Store and display custom characters                                         */
/*                                                                                                    */
/******************************************************************************************************/

class ParallaxLCD : public SoftwareSerial {
public:
   ParallaxLCD ( int pin, int numRows, int numCols );
   void setup ( boolean startEmpty=true ); 

   void on ();
   void off ();
   void lf ();
   void cr ();
   void empty ();

   void scrollLeft ();
   void scrollRight ();

   void backLightOn (); //Backlight only available with model numbers 27977, 27979
   void backLightOff (); //Backlight only available with model numbers 27977, 27979
   
   void pos ( int row, int col );

   void cursorUnderline();
   void cursorBlock();
   void cursorOnChrBlink();
   void cursorOff ();

   // Make music
   void playTone ( int duration, int scale, int note );
   
   // Custom characters
   void setCustomCharacter(int charIndex, unsigned char bytes[]);
   void printCustomCharacter(int charIndex);
   
	// shortcuts for printing at particular positions
   void at ( int row, int col, char );
   void at ( int row, int col, const char[] );
   void at ( int row, int col, String );
   void at ( int row, int col, uint8_t );
   void at ( int row, int col, int );
   void at ( int row, int col, unsigned int );
   void at ( int row, int col, long );
   void at ( int row, int col, unsigned long );
   void at ( int row, int col, long, int );

private:
	int _bv[9];
};


#endif


src/ParallaxLCD.cpp
/* 
	NOTE: you must: #include <SoftwareSerial.h>
	BEFORE including the class header file
	
	For Parallax LCD's (27976, 27977, 27979)
	For more information contact http://www.parallax.com	
	
	Updated to Arduino 1.0 by Chad Johnston (cjohnston@megatome.com)
	July 11, 2012
	      
	Original Work done by:
    Douglas Farrell	
    cyberblob59@yahoo.com
			  
    Used/Modified with permission from allen joslin
    --Based on SparkFunSerLCD.h from--
	allen joslin
	payson productions
	allen@joslin.net

	Updated 8/1/2012 Martin C. Stoufer mcstoufer@speakeasy.net
        # Added interface functions to play music and define/print custom characters
*/
/*
The MIT License (MIT)

Copyright (c) 2012-2018 Megatome Technologies, LLC

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/
#include <Arduino.h>

#include <SoftwareSerial.h>
#include "ParallaxLCD.h"

#define PINOUT      0
#define BOUNCE      1
#define NUMROWS     2
#define NUMCOLS     3

#define WRITE_BASE 	128
// The cursor position for 16 column and 20 column displays uses the same address
// The 16 column display just doesn't show the last 4 columns
#define COL_SIZE    20

ParallaxLCD::ParallaxLCD ( int pin, int numRows, int numCols) : SoftwareSerial(pin,pin) {
	_bv[PINOUT]=pin;
	_bv[BOUNCE]=10;
	_bv[NUMROWS]=numRows;
	_bv[NUMCOLS]=numCols;
}

void ParallaxLCD::setup(boolean startEmpty ) {
	pinMode(_bv[PINOUT], OUTPUT);
	delay(_bv[BOUNCE]);
	begin(19200);
	delay(_bv[BOUNCE]);
	if (startEmpty) {
		empty(); 
	}
	cursorOff();
}

void ParallaxLCD::on () { 
	write(24); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::off () { 
	write(21); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::scrollLeft () { 
	print(8); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::scrollRight () { 
	write(9); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::empty () { 
	write(12); delay(_bv[BOUNCE]*10);
}

void ParallaxLCD::cursorUnderline () { 
	write(24); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::cursorBlock () { 
	write(23); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::cursorOff () { 
	write(22); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::cursorOnChrBlink () { 
	write(25); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::cr () { 
	write(13); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::lf () { 
	write(10); delay(_bv[BOUNCE]); 
}

void ParallaxLCD::backLightOn () {
	write(17); delay(_bv[BOUNCE]);
}

void ParallaxLCD::backLightOff () {
	write(18); delay(_bv[BOUNCE]);
}

/**
 * Duration: 208 - 214 (1/64th note - 1 whole note[2 secs])
 * Scale   : 215 - 219 (3rd - 7th scale)
 * Note    : 220 - 232 (A, A#, B, etc..., F#, F, G#)
 */
void ParallaxLCD::playTone(int duration, int scale, int note) {
	
	duration = constrain(duration, 208, 214);
	scale = constrain(scale, 215, 219);
	note = constrain(note, 220, 232);

    write(scale);
    write(note);
    write(duration);
    delay(_bv[BOUNCE]);
}

/**
 * Set a custom character at the 'charIndex' index (0-7) with an
 *  array of bytes. Nominally, this will contain 8 bytes in it.
 */
void ParallaxLCD::setCustomCharacter(int charIndex, byte bytes[]) {

   if(sizeof(bytes)/sizeof(bytes[0]) > 8) {
      return;
   }

   int index = constrain(charIndex, 0, 7);
   index = map(index, 0, 7, 248, 255);
   
   unsigned char *byte = bytes;
   
   write(index);
   for (int i=0; i < 8; i++) {
     write(byte[i]);
   }
   delay(_bv[BOUNCE]);
}

/**
 * Print whatever is stored in the custom character buffer.
 * If nothing is stored there, usually, nothing will print out.
 */
void ParallaxLCD::printCustomCharacter(int charIndex) {
   int index = constrain(charIndex, 0, 7);
   write(index);
   delay(_bv[BOUNCE]);
}

void ParallaxLCD::pos ( int row, int col ) 
{
	write(WRITE_BASE + (row * COL_SIZE) + col);
}

// shortcuts

void ParallaxLCD::at ( int row, int col, char v )				{ pos(row,col); print(v); }
void ParallaxLCD::at ( int row, int col, const char v[] )	{ pos(row,col); print(v); }
void ParallaxLCD::at ( int row, int col, String v )             { pos(row,col); print(v); }
void ParallaxLCD::at ( int row, int col, uint8_t v )			{ pos(row,col); print(v); }
void ParallaxLCD::at ( int row, int col, int v )				{ pos(row,col); print(v); }
void ParallaxLCD::at ( int row, int col, unsigned int v )	{ pos(row,col); print(v); }
void ParallaxLCD::at ( int row, int col, long v )				{ pos(row,col); print(v); }
void ParallaxLCD::at ( int row, int col, unsigned long v )	{ pos(row,col); print(v); }
void ParallaxLCD::at ( int row, int col, long v, int t )		{ pos(row,col); print(v,t); }